server/src/main/java/org/apache/druid/server/coordination/ChangeRequestHttpSyncer.java

```diff
@@ -157,56 +176,85 @@ public void stop()
+   * Waits upto 1 milliseconds for the first successful sync with this server.
    */
-  public Map<String, Object> getDebugInfo()
+  public boolean isInitialized() throws InterruptedException
   {
-    long currTime = System.currentTimeMillis();
+    return initializationLatch.await(1, TimeUnit.MILLISECONDS);
+  }
 
-    Object notSuccessfullySyncedFor;
-    if (lastSuccessfulSyncTime == 0) {
-      notSuccessfullySyncedFor = "Never Successfully Synced";
-    } else {
-      notSuccessfullySyncedFor = (currTime - lastSuccessfulSyncTime) / 1000;
-    }
+  /**
+   * Returns debugging information for printing, must not be used for any other purpose.
+   */
+  public Map<String, Object> getDebugInfo()
+  {
     return ImmutableMap.of(
-        "notSyncedForSecs", lastSyncTime == 0 ? "Never Synced" : (currTime - lastSyncTime) / 1000,
-        "notSuccessfullySyncedFor", notSuccessfullySyncedFor,
-        "consecutiveFailedAttemptCount", consecutiveFailedAttemptCount,
+        "millisSinceLastSync",
+        sinceLastSyncRequest.isRunning() ? DateTimes.millisElapsed(sinceLastSyncRequest) : "Never synced",
+        "millisSinceLastSuccess",
+        sinceLastSyncSuccess.isRunning() ? DateTimes.millisElapsed(sinceLastSyncSuccess) : "Never synced successfully",
+        "millisUnstableDuration",
+        sinceUnstable.isRunning() ? DateTimes.millisElapsed(sinceUnstable) : "Stable",
+        "numRecentFailures", numRecentFailures,
         "syncScheduled", startStopLock.isStarted()
     );
   }
 
+  private boolean hasSyncedSuccessfullyOnce()
+  {
+    return initializationLatch.getCount() <= 0;
+  }
+
   /**
-   * Exposed for monitoring use to see if sync is working fine and not stopped due to any coding bugs. If this
-   * ever returns false then caller of this method must create an alert and it should be looked into for any
-   * bugs.
+   * Whether this syncer should be reset. This method returning true typically
+   * indicates a problem with the sync scheduler.
+   *
+   * @return true if the delay since the last sync request sent to the server
+   * has exceeded {@link #maxDelayBetweenSyncRequests}, false otherwise.
    */
-  public boolean isOK()
+  public boolean needsReset()
   {
-    return (System.currentTimeMillis() - lastSyncTime) < MAX_RETRY_BACKOFF + 3 * serverHttpTimeout;
+    return DateTimes.hasElapsed(maxDelayBetweenSyncRequests, sinceLastSyncRequest);
+  }
+
+  /**
+   * @return true if there have been no sync failures recently and the last
+   * successful sync was not more than {@link #maxMillisToWaitForSync} ago.
+   */
+  public boolean isSyncedSuccessfully()
+  {
+    final Duration timeoutDuration = Duration.millis(maxMillisToWaitForSync);
+    if (numRecentFailures > 0) {
+      return false;
+    } else if (hasSyncedSuccessfullyOnce()) {
+      return DateTimes.hasNotElapsed(timeoutDuration, sinceLastSyncSuccess);
+    } else {
+      return DateTimes.hasNotElapsed(timeoutDuration, sinceSyncerStart);
```

Comment:
Shouldn't we return false in this case?

Reply:
No, this statement says that we should wait for the timeoutDuration after startup before we make up our mind of syncer health. Until then, we just assume that the syncer is healthy. Otherwise, we would have a lot of unhealthy metrics right at startup, when in fact the initial delay is expected.

Reply:
Okay, got confused by the method name, which seems to be suggesting that we have synced successfully at least once.

Reply:
Yeah, good point. I will rename the method.